### ISSUE TITLE
Feature request: Support for Amplify manual deployments

### ISSUE THREAD

**Issue by draoncc** [hashicorp/terraform-provider-aws#24720]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS Amplify offers the option to do manual deployments through their API with the [CreateDeployment](https://docs.aws.amazon.com/amplify/latest/APIReference/API_CreateDeployment.html) and [StartDeployment](https://docs.aws.amazon.com/amplify/latest/APIReference/API_StartDeployment.html) actions. A resource like `aws_amplify_deployment` could be added for purposing these to support manual deployments in Amplify.

### New or Affected Resource(s)

* aws_amplify_deployment
* aws_amplify_branch (?)
* aws_amplify_app (?)

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
resource "aws_amplify_app" "example" {
  name = "app"
}

resource "aws_amplify_branch" "master" {
  app_id = aws_amplify_app.example.id
  branch_name = "master"
}

resource "aws_amplify_deployment" {
  app_id = aws_amplify_app.example.id
  branch_name = aws_amplify_branch.master.branch_name

  source_url = "https://s3.eu-central-1.amazonaws.com/my-s3-bucket/master.zip"
}
```

**Comment by chris-callahan** [comment]
bump

**Comment by mfedak-tqgg** [comment]
Great ides @draoncc I would really like to use this feature.

**Comment by VladyslavKurmaz** [comment]
extremely needed

**Comment by Sofranac-Boro** [comment]
bump

**Comment by advaithhl** [comment]
I'm surprised this doesn't exist till now! Bummer :(

**Comment by seamusriordan** [comment]
I have need for this

**Comment by rgant** [comment]
Based on the [AWS CLI documentation](https://docs.aws.amazon.com/amplify/latest/userguide/update-website-deployed-from-s3.html) I have a work-around until manual deploys are supported:

```tf
resource "null_resource" "deploy_from_s3" {
  triggers = {
    file_hashes = jsonencode({
      for fn in fileset("${path.root}/dist", "**") :
      fn => filesha256("${path.root}/dist/${fn}")
    })
  }

  provisioner "local-exec" {
    command = <<EOF
aws --profile personal --region us-east-1 \
  amplify start-deployment \
  --app-id ${module.amplify_website.amplify_site.id} \
  --branch-name ${module.amplify_website.amplify_site.branch_name} \
  --source-url s3://${aws_s3_bucket.website.id}/ \
  --source-url-type BUCKET_PREFIX
EOF
  }
}
```

[Full details](https://github.com/rgant/rob.gant.ninja/blob/7c6e98a99cd9f8668d7a6e7fee3ec7af7080662a/infrastructure/amplify.tf)

**Comment by cirnatdan** [comment]
What's the use case for this if created deployments expire after 8 hours?

**Comment by cirnatdan** [comment]
> [@cirnatdan](https://github.com/cirnatdan) what do you mean? A deployment is a process. When it is over the site is deployed forever.

Yes... I meant you have 8 hours to call `StartDeployment` and it cannot be used again.[1] What would be the utility in creating a deployment with terraform? Would you create a new terraform resource everytime there is a new change to the amplify app?

[1] https://docs.aws.amazon.com/amplify/latest/APIReference/API_CreateDeployment.html#API_CreateDeployment_RequestSyntax

**Comment by rgant** [comment]
When `source_url` changes, trigger a deployment. That doesn't need be a resource, it just needs to be an action.
